Ensure `@apply` works inside `@utility`

RobinMalfait opened the pull request:
This PR fixes an issue where `@apply` was not handled if it was used inside of `@utility`.

You should now be able to do something like this:
```css
@utility btn {
  @apply flex flex-col bg-white p-4 rounded-lg shadow-md;
}
```

If you then use `btn` as a class, the following CSS will be generated:
```css
.btn {
  border-radius: var(--radius-lg, .5rem);
  background-color: var(--color-white, #fff);
  padding: var(--spacing-4, 1rem);
  --tw-shadow: 0 4px 6px -1px #0000001a, 0 2px 4px -2px #0000001a;
  --tw-shadow-colored: 0 4px 6px -1px var(--tw-shadow-color), 0 2px 4px -2px var(--tw-shadow-color);
  box-shadow: var(--tw-inset-shadow), var(--tw-inset-ring-shadow), var(--tw-ring-offset-shadow), var(--tw-ring-shadow), var(--tw-shadow);
  flex-direction: column;
  display: flex;
}
```

This PR also makes sure that you can use custom `@utility` inside other `@utility` via `@apply`. E.g.:
```css
@utility foo {
  color: red;
}

@utility bar {
  color: red;
  @apply hover:foo;
}
```

If we detect a circular dependency, then we will throw an error since circular dependencies are not allowed. E.g.:
```css
@utility foo {
  @apply hover:bar;
}

@utility bar {
  @apply focus:baz;
}

@utility baz {
  @apply dark:foo;
}
```
Regardless of which utility you use, eventually it will apply itself.

Fixes: #14143

RobinMalfait committed:
fix typo

RobinMalfait committed:
move handling of `@apply` to dedicated function

RobinMalfait committed:
handle `@apply` inside of `@utility`

RobinMalfait committed:
use better name

RobinMalfait committed:
add nested example

RobinMalfait committed:
add test to detect circular dependencies

thecrypticace commented on packages/tailwindcss/src/utilities.test.ts:
```diff
@@ -15149,4 +15149,66 @@ describe('custom utilities', () => {
+    `)
+  })
+
+  test('custom utilities with `@apply` causing circular dependencies should error', () => {
```
We should probably add a test verifying that apply of other `@utility`s work in an `@utility` — not just for the circular dep error case.

RobinMalfait committed:
add test to verify that you can reference another `@utility` via `@apply`

adamwathan commented on packages/tailwindcss/src/index.ts:
```diff
@@ -113,7 +113,11 @@ export function compile(
-          return structuredClone(node.nodes)
+          let ast = structuredClone(node.nodes)
+
+          substituteAtApply(ast, designSystem)
```
Feels a bit weird to me to make this the responsibility of the actual plugin/matcher/whatever 🤔

Should we maybe just handle all of the `@apply` stuff in a separate pass at the end? Or perhaps we could replace it in the CSS up front, so we take this:

```css
@utility foo {
  @apply flex
}
```

...and turn that into:

```css
@utility foo {
  & {
    display: flex
  }
}
```

...before the plugin is even registered with the design system?

thecrypticace replied:
I think this would cause problems when a custom utility is used in an `@apply` inside another custom utility.

thecrypticace replied:
Doing this as a separate pass seems reasonable. We could probably:
- Perform a single pass in `compile()` — making sure there's less to replace later
- Perform a pass every time we call `build()` — and we'd only ever need to do this for the _new_ nodes.

RobinMalfait replied:
I 100% agree, but these are also lazily evaluated, so it will only take effect the moment we actually use the utility. If you never use them in the first place, handling the `@apply` is a bit wasteful.

> Replace `@apply` in a separate pass at the end

That's an option, but as long as the CSS files don't change, the `@apply` directives used in the CSS itself won't change either. So doing this as a separate pass whenever we build the final CSS based on the candidates seems a bit wasteful. (Because then we would do it `n` times for every `build()` instead of once)

The `@apply` used in a `@utility` should only be handled if the utility is being used somewhere, so this is only available once we do the `build(candidates)` call.

> Or perhaps we could replace it in the CSS up front

We can do that for _some_ utilities (the built-in ones, and custom `@utility` ones that were registered before we see the first `@apply`). If you have this setup:

```css
@utility foo {
  @apply hover:bar;
}

@utility bar {
  color: red;
}
```

Then we can't replace the `@apply hover:bar` in `foo` yet because the `bar` utility has not been registered yet.

Another issue is that we won't be able to detect circular references this way. If `bar` happens to reference `foo` then we should detect that.

RobinMalfait replied:
We can however collect the nodes and once all custom `@utility` utilities are registered, replace the `@apply` once and later we can re-use the replaced `@apply` nodes.

RobinMalfait replied:
I think I figured out a nice way of handling this (https://github.com/tailwindlabs/tailwindcss/pull/14144/commits/48ff0b8884e23b546f377fc4c10ee312166bffb6)

1. Collect the `@utility` rules _if_ they contain `@apply`
2. Register the `@utility` in the design sytem
3. Once all `@utility` utilities are registered, substitute the `@apply` with its correct rule. This way we don't have to handle the `@apply` over and over and over again, just once.
4. Bonus points, since we register all utilities first, we can apply `@utility` out of order and detect circular dependencies.

RobinMalfait committed:
add test to verify the order of `@utility` and `@apply` doesn't matter

RobinMalfait committed:
adjust comment

We are not replace it yet, so let's make that clear

RobinMalfait committed:
use early return

This allows us to dedent a level of nesting.

RobinMalfait committed:
detect circular dependencies

RobinMalfait committed:
only substitute the `@apply` rules when they are used

The `substituteAtApply()` should only be executed _if_ we are using
`@apply`, we don't have to execute this for every single `@utility`,
especially if they don't even use `@apply`.

RobinMalfait committed:
rename variable

Not 100% convinced of the naming, but feels better than `seen` 🤔

RobinMalfait committed:
make new tests async

RobinMalfait committed:
handle `@apply` once

This now collects the custom `@utility` rules if they contain `@apply`.
Once all the `@utility` utilities are registered in the system, then we
will update (mutate) the `node` to replace all the `@apply` rules with
its substitute.

This means that by the time you are using the `@utility`, the `@apply`
is already handled and we don't pay any penalties that require handling
the `@apply`.

We wait for all `@utility` utilities to be registered, so that we can
use one `@utility` via `@apply` in another (out of order).

We will also verify that we don't run into circular dependencies
upfront.

RobinMalfait commented on packages/tailwindcss/src/index.ts:
```diff
@@ -352,41 +366,14 @@ export async function compile(
-      }
-    })
+    if (customUtilitiesWithAtApply.length > 0) {
+      substituteAtApply(customUtilitiesWithAtApply.splice(0), designSystem)
```
Using a `.splice(0)` to empty out the array because we don't need to hold onto those nodes anymore.

RobinMalfait commented on packages/tailwindcss/src/index.ts:
```diff
@@ -439,6 +426,80 @@ export async function compile(
+        //
+        // Figure out which candidate caused the circular dependency. This will
+        // help to create a useful error message for the end user.
+        let candidate = candidates.find((candidate) => {
+          let selector = `.${escape(candidate)}`
+          let found = false
+
+          for (let rule of candidateAst) {
+            if (rule.kind !== 'rule') continue
+            if (rule.selector !== selector) continue
+
+            walk(rule.nodes, (child) => {
+              if (child === node) {
+                found = true
+                return WalkAction.Stop
+              }
+            })
+
+            if (found) return found
+          }
+
+          return found
+        })
```
We are walking inside of a `.find()` which could be slow. BUT, this only happens if we already know that we have a circular dependency so we will throw an error anyway. We are just doing this to figure out which candidate the culprit is to improve the error message.

RobinMalfait commented on packages/tailwindcss/src/index.ts:
```diff
@@ -110,6 +112,18 @@ export async function compile(
+
+        customUtilitiesWithAtApply.push(node)
+
+        return WalkAction.Stop
```
Once we know `@apply` is being used, we can stop early. We will handle all the `@apply` nodes later.

RobinMalfait committed:
update changelog

RobinMalfait commented on packages/tailwindcss/src/index.ts:
```diff
@@ -439,6 +426,80 @@ export async function compile(
+
+      // Verify that we don't have any circular dependencies by verifying that
+      // the current node does not appear in the new nodes.
+      walk(newNodes, (child) => {
```
This function is a copy of the existing `@apply` handling.

The only differences are:
1. This additional walk to detect circular dependencies
2. The early returns at the top to reduce the level of nesting

RobinMalfait committed:
replace `@utility` after we handled `@apply`

We can use `@apply` inside of `@utility` calls. Before this we collected
the `@utility` before replacing it with `[]`.

Instead, we don't mutate the AST yet, handle `@apply` once, last we
remove the `@utility` at-rules. Because we know that `@utility` has to
be top-level, we can also improve the performance of the `walk` by never
traversing into nested nodes.

RobinMalfait commented on packages/tailwindcss/src/index.ts:
```diff
@@ -353,42 +353,23 @@ export async function compile(
-        }
-      }
-    })
+    substituteAtApply(ast, designSystem)
```
We could technically re-inline this function. But we will have to split it up immediately once this is merged because it's necessary in this PR: #14114

RobinMalfait committed:
throw error as soon as possible

Instead of cleanly wrapping up the `walk()`, and tracking whether we
found the candidate or not. We can just throw immediately once we found
the culprit.

RobinMalfait committed:
add test to verify that circular dependencies are detected when nesting `@apply`
